File: src/tools/pmd-cpd/eval-repos/synthetic/typescript/light_duplication.ts
/**
 * TypeScript file with light duplication - small duplicated blocks.
 */

interface UserData {
    id?: number;
    name?: string;
    email?: string;
    active?: boolean;
    createdAt?: string;
    permissions?: string[];
}

function processUserData(user: UserData): UserData {
    return {
        id: user.id || 0,
        name: (user.name || "").trim(),
        email: (user.email || "").toLowerCase().trim(),
        active: user.active !== false,
        createdAt: user.createdAt || "",
    };
}

function processAdminData(admin: UserData): UserData {
    return {
        id: admin.id || 0,
        name: (admin.name || "").trim(),
        email: (admin.email || "").toLowerCase().trim(),
        active: admin.active !== false,
        createdAt: admin.createdAt || "",
        permissions: admin.permissions || [],
    };
}

function validateEmail(email: string): boolean {
    if (!email) return false;
    if (!email.includes("@")) return false;
    const parts = email.split("@");
    if (parts.length !== 2) return false;
    if (!parts[0] || !parts[1]) return false;
    if (!parts[1].includes(".")) return false;
    return true;
}

function formatCurrency(amount: number, currency: string = "USD"): string {
    const symbols: Record<string, string> = { USD: "$", EUR: "E", GBP: "P" };
    const symbol = symbols[currency] || currency;
    if (amount < 0) {
        return `-${symbol}${Math.abs(amount).toFixed(2)}`;
    }
    return `${symbol}${amount.toFixed(2)}`;
}

function calculateDiscount(price: number, discountPct: number): number {
    if (discountPct < 0 || discountPct > 100) {
        throw new Error("Discount must be between 0 and 100");
    }
    return price * (1 - discountPct / 100);
}

export { UserData, processUserData, processAdminData, validateEmail, formatCurrency, calculateDiscount };

File: src/tools/pmd-cpd/eval-repos/synthetic/csharp/LightDuplication.cs
/// <summary>
/// C# file with light duplication - small duplicated blocks.
/// </summary>

using System;
using System.Collections.Generic;

namespace Synthetic.LightDuplication
{
    public class UserData
    {
        public int Id { get; set; }
        public string Name { get; set; }
        public string Email { get; set; }
        public bool Active { get; set; } = true;
        public string CreatedAt { get; set; }
        public List<string> Permissions { get; set; }
    }

    public static class DataProcessor
    {
        public static UserData ProcessUserData(UserData user)
        {
            return new UserData
            {
                Id = user.Id,
                Name = (user.Name ?? "").Trim(),
                Email = (user.Email ?? "").ToLower().Trim(),
                Active = user.Active,
                CreatedAt = user.CreatedAt ?? ""
            };
        }

        public static UserData ProcessAdminData(UserData admin)
        {
            return new UserData
            {
                Id = admin.Id,
                Name = (admin.Name ?? "").Trim(),
                Email = (admin.Email ?? "").ToLower().Trim(),
                Active = admin.Active,
                CreatedAt = admin.CreatedAt ?? "",
                Permissions = admin.Permissions ?? new List<string>()
            };
        }

        public static bool ValidateEmail(string email)
        {
            if (string.IsNullOrEmpty(email)) return false;
            if (!email.Contains("@")) return false;
            var parts = email.Split('@');
            if (parts.Length != 2) return false;
            if (string.IsNullOrEmpty(parts[0]) || string.IsNullOrEmpty(parts[1])) return false;
            if (!parts[1].Contains(".")) return false;
            return true;
        }

        public static string FormatCurrency(decimal amount, string currency = "USD")
        {
            var symbols = new Dictionary<string, string>
            {
                { "USD", "$" },
                { "EUR", "E" },
                { "GBP", "P" }
            };
            var symbol = symbols.GetValueOrDefault(currency, currency);
            if (amount < 0)
            {
                return $"-{symbol}{Math.Abs(amount):N2}";
            }
            return $"{symbol}{amount:N2}";
        }

        public static decimal CalculateDiscount(decimal price, decimal discountPct)
        {
            if (discountPct < 0 || discountPct > 100)
            {
                throw new ArgumentException("Discount must be between 0 and 100");
            }
            return price * (1 - discountPct / 100);
        }
    }
}
